### DISCUSSION TITLE
Does "<" work with this version?

### DISCUSSION THREAD

**Discussion by Mistborn-First-Era** [AlUlkesh/stable-diffusion-webui-images-browser#59] [Q&A] [answered]
I just started digging a bit and realized there was a new fork. 
In the old original version "<" wouldn't be accepted and therefore I couldn't make a Lora wildcard inspiration. Has this been fixed in the latest version? I wanted to check before I updated and moved everything around.

**Comment by AlUlkesh** [comment]
Accepted where exactly? reading exif? Search? Custom paths?

**Comment by Mistborn-First-Era** [comment]
I made a .txt file with line seperated values such as 

hairstyles.txt
```
bangs
top knot
pony tail
<lora:PinappleOnHeadHair:1>
```

It would error out when I used hairstyles.txt for the inspiration images script in A1111.

Hopefully that explains it better.

**Reply by AlUlkesh** [reply]
I just tried it out. It breaks here:
```
  File "C:\tools\Ai\stable-diffusion-webui\extensions\stable-diffusion-webui-inspiration\scripts\inspiration.py", line 59, in run
    os.mkdir(artist_path)
```
That's inside the inspiration extension, not the images browser.

The fix would be very minor, but yfszzx hasn't been seen since December.

I wonder if I should make a fork of that one too. You're obviously using that one.

**Reply by Mistborn-First-Era** [reply]
Ah sorry, I thought this was the inspiration extension fork. I guess Image browser is a fork of inspiration? 
If it's a simple fix I would really appreciate it.
I looked into it with my limited python ability. 
It looks like the problem is that it makes sub folders and then uses those folder names as prompts.

I did some debugging and the there are actually two problems that can probably be addressed at the same time.
1st: Since file names cannot have certain symbols 
  ```
< (less than)
  > (greater than)
  : (colon - sometimes works, but is actually NTFS Alternate Data Streams)
  " (double quote)
  / (forward slash)
  \ (backslash)
  | (vertical bar or pipe)
  ? (question mark)
  * (asterisk)
```
Those cannot be used in your inspiration txt file. which is pretty limiting. I don't know if their would be a conflict in the ui as well within the "Inspiration tab -> get inspiration -> right side under get inspiration" since that uses the file name to allow for easy copy pasting.

2nd: This is probably a problem that originates from the 1st problem the more I think about it. But you cannot have "," (a comma) in your 'inspiration' prompt. For example
```
dog
cat
mouse
horse, farm, springtime
tiger, <lora:tigerlora:1>
```
would make the folders dog, cat, mouse, horse, and tiger; within each of which it would use only those words to generate the prompt. It cuts off farm, springtime, and <<lora:tigerlora:1>> from their respective lines.

Seems like it may be annoying to fix. I am pretty new to python still so I don't think I could fix it myself, but I think I did a good job debugging it.

**Reply by AlUlkesh** [reply]
Yes, I see. Well you could solve them both the same way by adding comma to that list of characters. And when you access paths you'd replace them by something else "_" for example. But keep the original data, so it can be used for the prompt and other things.

I'll look into it.

**Comment by Mistborn-First-Era** [comment, marked as answer]
I did it yay. 

This is the code I have in the run function within the main class inside root->extensions->inspiration->scripts

It takes uses the first word of the inspiration.txt file as the folder name (it removes any illegal symbols and changes them to _)
then it takes the full prompt and uses that to generate the inspiration images.

for example:
```
dog
cat, woman, balcony
<lora:lion:1>, jungle, masterpiece
?????????pie???????????, curious
```

would make the folders
```
dog
cat
_lora_lion_1_
_________pie_________
```

Should work as the new run command.

```
       def run(self, p, prompt_placeholder, files):
        if not re.search(prompt_placeholder, p.prompt):
            raise Exception("Missing " + prompt_placeholder + " in prompt")
        original_prompt = p.prompt
        p.do_not_save_samples = True
        p.do_not_save_grid = True
        for file in files:
            tp = file.orig_name.split(".")[0]
            print("Process " + tp + "..")
            path = os.path.join(inspiration_dir, tp)
            if not os.path.exists(path):
                os.makedirs(path)
            f = open(file.name, "r", encoding='utf-8')
            line = f.readline()
            while len(line) > 0:
                line = line.rstrip("\n")
                # Ignore headline of artist.csv from automatic1111 repository
                if line == "artist,score,category":
                    line = f.readline()
                    continue
                double_quotes_match = re.search('^".+"', line)
                if double_quotes_match:
                    fixed_name = double_quotes_match.group()[1:len(double_quotes_match.group()) - 1]
                else:
                    name = line.split(",")[0]
                    fixed_name = ""
                    for letter in name:
                        bad_symbol_list = ["<", ">", "?", "\\", "*", ":", '"', "'", "/", "|", "?"]
                        if letter in bad_symbol_list:
                            fixed_name += "_"
                        else:
                            fixed_name += letter
                artist_path = os.path.join(path, fixed_name)
                if not os.path.exists(artist_path):
                    os.mkdir(artist_path)
                if len(os.listdir(artist_path)) >= opts.inspiration_max_samples:
                    print("Limit of " + str(opts.inspiration_max_samples) + " of '" + fixed_name + "' in " + tp + " is reached. Limit can be changed in the settings.")
                    continue
                p.prompt = re.sub(prompt_placeholder, line, original_prompt)
                print("folder name is", fixed_name, "using", p.prompt, "as the prompt.")
                line = f.readline()
                processed = processing.process_images(p)
                for img in processed.images:
                    i = 0
                    filename = os.path.join(artist_path, format(0, "03d") + ".jpg")
                    while os.path.exists(filename):
                        i += 1
                        filename = os.path.join(artist_path, format(i, "03d") + ".jpg")
                    img.save(filename, quality=80)
        return processed
```
ps. how do I fork myself? I would like to get a green dot on my github record I can use to help land a job once I am done with school.

**Reply by AlUlkesh** [reply]
Nice :smiley:

You can fork here, top of the page:
![image](https://user-images.githubusercontent.com/99896447/222679324-f754d83e-3f61-440b-9d79-fdb5d87e50ad.png)

Small tip: regex is really good for replacing those values, for example:
```
invalid_chars = '<>:\"/\\\\|?*\0'
fixed_name = re.sub(f'[{re.escape(invalid_chars)}]', '_', name)
```

**Comment by Mistborn-First-Era** [comment]
Think I did it. Thanks for the help. I will have to learn regex. I tried, but it was hard to understand.
https://github.com/Mistborn-First-Era/stable-diffusion-webui-inspiration-Functioning-Loras
